sympy/matrices/graph.py

```diff
@@ -0,0 +1,135 @@
+from sympy.utilities.iterables import \
+    flatten, strongly_connected_components
+
+from .common import NonSquareMatrixError
+
+
+def _connected_components(M):
+    """Returns the list of connected vertices of the graph when
+    a square matrix is viewed as a weighted graph.
+
+    Examples
+    ========
+
+    >>> from sympy import symbols, Matrix
+    >>> a, b, c, d, e, f, g, h = symbols('a:h')
+    >>> A = Matrix([
+    ...     [a, 0, b, 0],
+    ...     [0, e, 0, f],
+    ...     [c, 0, d, 0],
+    ...     [0, g, 0, h]])
+    >>> A.connected_components()
+    [[0, 2], [1, 3]]
+
+    Notes
+    =====
+
+    Even if any symbolic elements of the matrix can be indeterminate
+    to be zero mathematically, this only takes the account of the
+    structural aspect of the matrix, so they will considered to be
+    nonzero.
+    """
+    from .sparse import SparseMatrix
+
+    if not M.is_square:
+        raise NonSquareMatrixError
+
+    N = M.shape[0]
+    V = list(range(N))
+
+    if isinstance(M, SparseMatrix):
+        E = set()
+        for i, j in M._smat.keys():
+            if i == j:
+                E.add((i, j))
+            else:
+                E.add((i, j))
+                E.add((j, i))
+        E = list(E)
+    else:
+        E = list()
+        for i in range(N):
+            for j in range(i):
+                if M[i, j] is M.zero and M[j, i] is M.zero:
+                    continue
+                E.append((i, j))
+                E.append((j, i))
+            if M[i, i] is M.zero:
+                E.append((i, i))
+
+    return strongly_connected_components((V, E))
```

Comment:
This should probably be `connected_components` in which case duplicate edges don't need to be added here.

Reply:
With todok this can just be:
```
return connected_components((range(N), M.todok()))
```

Reply:
Now, the test fails in python3.5 probably because of dictionary ordering. Does `connected_components` make unique results? If not, I'm thinking of sorting the output.

Reply:
The results returned from connected_components can be in any order. The strongly_connected_components function returns components in topological order which is also not unique.